src/pgen/kh.cpp

```diff
@@ -32,61 +32,32 @@
 
 // AthenaPK headers
 #include "../main.hpp"
+#include "utils/error_checking.hpp"
 
 namespace kh {
 using namespace parthenon::driver::prelude;
 
 //----------------------------------------------------------------------------------------
-//! \fn void MeshBlock::ProblemGenerator(ParameterInput *pin)
+//! \fn void Mesh::ProblemGenerator(Mesh *pm, ParameterInput *pin, MeshData<Real> *md)
 //  \brief Problem Generator for the Kelvin-Helmholtz test
 
-void ProblemGenerator(MeshBlock *pmb, ParameterInput *pin) {
+void ProblemGenerator(Mesh *pmesh, ParameterInput *pin, MeshData<Real> *md) {
   auto vflow = pin->GetReal("problem/kh", "vflow");
   auto iprob = pin->GetInteger("problem/kh", "iprob");
+  // Get pointer to first block (always exists) for common data like loop bounds
+  auto pmb = md->GetBlockData(0)->GetBlockPointer();
   auto ib = pmb->cellbounds.GetBoundsI(IndexDomain::interior);
   auto jb = pmb->cellbounds.GetBoundsJ(IndexDomain::interior);
   auto kb = pmb->cellbounds.GetBoundsK(IndexDomain::interior);
   auto gam = pin->GetReal("hydro", "gamma");
   auto gm1 = (gam - 1.0);
 
-  // initialize conserved variables
-  auto &rc = pmb->meshblock_data.Get();
-  auto &u_dev = rc->Get("cons").data;
-  auto &coords = pmb->coords;
-  // initializing on host
-  auto u = u_dev.GetHostMirrorAndCopy();
+  // Initialize conserved variables
+  // Get a MeshBlockPack on device with all conserved variables
+  const auto &cons = md->PackVariables(std::vector<std::string>{"cons"});
+  const auto num_blocks = md->NumBlocks();
 
-  std::mt19937 gen(pmb->gid); // Standard mersenne_twister_engine seeded with gid
-  std::uniform_real_distribution<Real> ran(-0.5, 0.5);
-
-  //--- iprob=1.  Uniform stream with density ratio "drat" located in region -1/4<y<1/4
-  // moving at (-vflow) seperated by two slip-surfaces from background medium with d=1
-  // moving at (+vflow), random perturbations.  This is the classic, unresolved K-H test.
-
-  if (iprob == 1) {
-    // Read problem parameters
-    Real drat = pin->GetReal("problem/kh", "drat");
-    Real amp = pin->GetReal("problem/kh", "amp");
-    for (int k = kb.s; k <= kb.e; k++) {
-      for (int j = jb.s; j <= jb.e; j++) {
-        for (int i = ib.s; i <= ib.e; i++) {
-          u(IDN, k, j, i) = 1.0;
-          u(IM1, k, j, i) = vflow + amp * ran(gen);
-          u(IM2, k, j, i) = amp * ran(gen);
-          u(IM3, k, j, i) = 0.0;
-          if (std::abs(coords.Xc<2>(j)) < 0.25) {
-            u(IDN, k, j, i) = drat;
-            u(IM1, k, j, i) = -drat * (vflow + amp * ran(gen));
-            u(IM2, k, j, i) = drat * amp * ran(gen);
-          }
-          // Pressure scaled to give a sound speed of 1 with gamma=1.4
-          u(IEN, k, j, i) =
-              2.5 / gm1 +
-              0.5 * (SQR(u(IM1, k, j, i)) + SQR(u(IM2, k, j, i))) / u(IDN, k, j, i);
-        }
-      }
-    }
-  }
+  //--- iprob=1. This was the classic, unresolved K-H test.
```

Comment:
is it worth throwing a special error for this?  i.e., "this version of the test is unresolved; try iprob >= 2"

Reply:
It's already caught at the bottom of the if statements, so any non-implemented `iprob` will raise an explicit error.